geode-core/src/main/java/org/apache/geode/distributed/internal/DistributionAdvisor.java

```diff
@@ -256,11 +258,23 @@ public void syncForCrashedMember(final InternalDistributedMember id, final Profi
 
     final boolean isDebugEnabled = logger.isDebugEnabled();
     if (isDebugEnabled) {
-      logger.debug("da.syncForCrashedMember will sync region in waiting thread pool: {}", dr);
+      logger.debug("da.syncForCrashedMember will sync region in cache's timer for region: {}", dr);
     }
-    dr.getDistributionManager().getWaitingThreadPool().execute(new Runnable() {
-      // bug #49601 - don't synchronize until GII has been performed
-      public void run() {
+    // schedule the synchronization for execution in the future based on the client health monitor
+    // interval. This allows client caches to retry an operation that might otherwise be recovered
+    // through the sync operation. Without associated event information this could cause the
+    // retried operation to be mishandled. See GEODE-5505
+    long delay;
+    try {
+      delay = dr.getGemFireCache().getCacheServers().stream().max(
+          (o1, o2) -> o1.getMaximumTimeBetweenPings() - o2.getMaximumTimeBetweenPings()).get()
```

Comment:
I'm new to the Stream API but I find this code a bit obscure. I also don't like the need to catch the exception. The three call of getMaximumTimeBetweenPings makes it very long and the third call is hard to understand initially.
After fumbling around for a while in my IDE I came up with this which I think is easier to understand. It compiles; I'm not sure it does all the right things at runtime:
    final long delay = dr.getGemFireCache().getCacheServers().stream().mapToLong(o -> o.getMaximumTimeBetweenPings()).max().orElse(0L);

Reply:
Your solution looks nice.  In my code I'm just using max(Comparator) to sort the cache servers.

Reply:
Something that might be worth cleaning up about this code is that it later does some checks and decides that no sync is needed. For example it does this one:
dr.getDataPolicy().withPersistence() && persistentId == null
and then in synchronizeForLostMember() it does nothing if we don't have concurrency checks enabled.
It seems like all these checks could be done synchronously before hand this off to the scheduler.
It also seems like getRegionForDeltaGII could be renamed (i.e. getRegionForLostMemberSynchronization) or turned into another precondition (that our region is a DistributedRegion). Better would be just to call a method on InternalRegion that does nothing on LocalRegion and is only implemented on DIstributedRegion instead of doing this instanceof check.

Reply:
Also in the run2() loop we only check dr.isDestroyed if we are not initialized. Shouldn't we check it once more after the loop to make sure we do the destroy check on an initialized region?